casper/src/main/scala/coop/rchain/casper/storage/LmdbStoreManager.scala

```diff
@@ -0,0 +1,138 @@
+package coop.rchain.casper.storage
+
+import java.nio.ByteBuffer
+import java.nio.file.{Files, Path}
+
+import cats.effect.concurrent.{Deferred, Ref}
+import cats.effect.{Concurrent, Sync}
+import cats.syntax.all._
+import coop.rchain.shared.Log
+import coop.rchain.store.{KeyValueStore, KeyValueStoreManager}
+import enumeratum.{Enum, EnumEntry}
+import org.lmdbjava.{DbiFlags, Env, EnvFlags}
+
+object LmdbStoreManager {
+  def apply[F[_]: Concurrent: Log](dirPath: Path): F[KeyValueStoreManager[F]] =
+    Deferred[F, Env[ByteBuffer]] map (LmdbStoreManagerImpl(dirPath, _))
+}
+
+private final case class LmdbStoreManagerImpl[F[_]: Concurrent: Log](
+    dirPath: Path,
+    envDefer: Deferred[F, Env[ByteBuffer]]
+) extends KeyValueStoreManager[F] {
+
+  sealed trait EnvRefStatus extends EnumEntry
+  object EnvRefStatus extends Enum[EnvRefStatus] {
+    val values = findValues
+    case object EnvClosed   extends EnvRefStatus
+    case object EnvStarting extends EnvRefStatus
+    case object EnvOpen     extends EnvRefStatus
+    case object EnvClosing  extends EnvRefStatus
+  }
+  import EnvRefStatus._
+
+  private case class DbState(
+      status: EnvRefStatus,
+      inProgress: Int,
+      envDefer: Deferred[F, Env[ByteBuffer]],
+      dbs: Map[String, Deferred[F, DbEnv[F]]] = Map.empty
+  ) {
+    override def toString() = s"DbState(status: $status, inProgress: $inProgress)"
+  }
+
+  // Internal manager state for LMDB databases and environments
+  private val varState = Ref.unsafe(DbState(EnvClosed, 0, envDefer))
+
+  override def database(dbName: String): F[KeyValueStore[F]] =
+    Log[F].info(s"LMDB database created path: $dirPath, DB: $dbName").pure[F] >>
+      Sync[F].delay(new LmdbKeyValueStore[F](getCurrentEnv(dbName)))
+
+  private def getCurrentEnv(dbName: String): F[DbEnv[F]] =
+    for {
+      // Check if environment is closed
+      isNewEnv <- varState.modify { st =>
+                   if (st.status == EnvClosed) {
+                     val newState = st.copy(status = EnvStarting)
+                     (newState, true)
+                   } else (st, false)
+                 }
+      // Create new environment when it's closed.
+      // IMPORTANT: this can happen only once, all other callers
+      // are blocked on Deferred object until it's completed.
+      _ <- createEnv.whenA(isNewEnv)
+
+      // Block until environment is ready.
+      st  <- varState.get
+      env <- st.envDefer.get
+
+      // Find DB ref / first time create deferred object.
+      newDbDefer <- Deferred[F, DbEnv[F]]
+      action <- varState.modify { st =>
+                 val maybeDbRef = st.dbs.get(dbName)
+                 maybeDbRef.fold {
+                   val newDbs = st.dbs.updated(dbName, newDbDefer)
+                   val newSt  = st.copy(dbs = newDbs)
+                   (newSt, (true, newDbDefer))
+                 }(defer => (st, (false, defer)))
+               }
```

Comment:
So each defer in each db can be a guarantee that the db can be used only if it is created, right?
It looks like the `true` `false` are dups? Not sure if i am right.

Reply:
@zsluedem yes, each caller will be blocked with Deferred until it's completed which is done when DB is created. This is the first step to implement reload of the environment where we need to wait operations on existing environment to finish.

Returning a flag if new environment should be created is a way to capture transition that must happen only once. Only the first request should trigger creation of a new database. But this "effect" cannot be executed inside Ref `modify` function because this function should only depend on the internal state and can be executed multiple times if the state is changed between reading and writing. So the flag `true` will be only signaled once per database for "effectful" code to execute.